src/shared/copy.c

```diff
@@ -711,7 +712,9 @@ static int fd_copy_regular(
                 if (r < 0)
                         return r;
         }
-        fdt = openat(dt, to, O_WRONLY|O_CREAT|O_EXCL|O_CLOEXEC|O_NOCTTY|O_NOFOLLOW, st->st_mode & 07777);
+        if (copy_flags & COPY_REPLACE)
+                dt_flags &= ~O_EXCL;
```

Comment:
this doesn't work. Consider an existing file that is large, and you copy a small file over. If you just drop O_EXCL, you'd leave the file at its original size, and just override the beginning of it, but that's not what you want here.

So, do what 'cp' does: open for writing with O_EXCL, and if this yields EEXIST, then unlink the file, and try again.

Reply:
also, this shouldn't just be done for regular files, but for any form of inodes we copy. i.e. the right place to put this unlink-and-retry logic is in fd_copy_directory() and copy_tree_at_full().

Reply:
Thanks for the catch!
Just so that I understand clearly, should we try to open the file and unlink if it returns EEXIST? Or is there an API we can use to check if the file exists (that is perhaps cheaper) and conditionally unlink?
Apologies if my question doesn't make sense, as I am new to using the Linux API

Reply:
yes, try to open it with O_EXCL, as before, but then if EEXIST is seen and the COPY_REPLACE flag is set, unlink the file, and do the same thing again. If that then still fails with EEXIST something is off, just propagate the EEXIST.

Reply:
Thanks again for the review.
I have tried to address this by refactoring the "leaf" nodes of the file tree into a function, which I can then call twice in the case of EEXIST.
This is so that when copying a directory, we can selectively "unlink and retry" for only "leaf" nodes (i.e. not directory nodes).
Would appreciate your thoughts on this when you have time :)